core/common/src/main/scala/org/typelevel/otel4s/Attributes.scala

```diff
@@ -93,43 +99,74 @@ sealed trait Attributes
+    * If multiple [[`Attribute`]]s in `this` and/or `that` have the same name,
     * only the final one (according to `that`'s iterator) will be retained in
     * the resulting `Attributes`.
     */
   def concat(that: IterableOnce[Attribute[_]]): Attributes =
-    attributesFactory.fromSpecific(this.view ++ that)
+    that match {
+      case other: Attributes =>
+        uncheckedFromMap(asUntypedMap ++ other.asUntypedMap)
+      case other =>
+        uncheckedFromMap(
+          asUntypedMap ++ other.iterator.map(a => a.key.name -> a)
+        )
+    }
 
   /** Invariant overload of [[scala.collection.IterableOps.++ `IterableOps#++`]]
     * that returns `Attributes` rather than `Iterable`.
     *
-    * If multiple [[`Attribute`]]s in `this` and/or `that` have the same key,
+    * If multiple [[`Attribute`]]s in `this` and/or `that` have the same name,
     * only the final one (according to `that`'s iterator) will be retained in
     * the resulting `Attributes`.
     */
   final def ++(that: IterableOnce[Attribute[_]]): Attributes =
     concat(that)
 
+  /** Removes all attributes with any of the given key names. */
+  def removedAllUntyped(names: IterableOnce[String]): Attributes =
+    uncheckedFromMap(asUntypedMap.removedAll(names))
+
   /** Removes all attributes with any of the given keys. */
-  def removedAll(that: IterableOnce[AttributeKey[_]]): Attributes =
-    attributesFactory.fromSpecific((toMap -- that).values)
+  def removedAll(keys: IterableOnce[AttributeKey[_]]): Attributes =
+    keys.iterator.foldLeft(this)(_ - _)
 
   /** Removes all attributes with any of the given keys. */
-  final def --(that: IterableOnce[AttributeKey[_]]): Attributes =
-    removedAll(that)
+  final def --(keys: IterableOnce[AttributeKey[_]]): Attributes =
+    removedAll(keys)
 
-  /** @return the `Map` representation of these `Attributes` */
-  def toMap: Map[AttributeKey[_], Attribute[_]]
+  /** @return
+    *   the `Map` representation of these `Attributes` with typed
+    *   [[`AttributeKey`]]s
+    */
+  def asMap: Map[AttributeKey[_], Attribute[_]] = new TypedMap
+
+  /** @return
+    *   the `Map` representation of these `Attributes` with untyped `String`
+    *   keys
+    */
+  def asUntypedMap: Map[String, Attribute[_]]
+
+  /** Equivalent to `asMap.keySet`.
+    *
+    * @return
+    *   the typed [[`AttributeKey` keys]] of the [[`Attribute`]]s
+    */
+  final def keys: Set[AttributeKey[_]] = asMap.keySet
 
-  /** Equivalent to `toMap.keySet`.
+  /** Equivalent to `asUntypedMap.keySet`.
     *
     * @return
-    *   the keys of the [[`Attribute`]]s
+    *   the untyped `String` key names of the [[`Attribute`]]s
     */
-  final def keys: Set[AttributeKey[_]] = toMap.keySet
+  final def untypedKeys: Set[String] = asUntypedMap.keySet
 
   /** A factory for creating `Attributes`. */
-  def attributesFactory: SpecificIterableFactory[Attribute[_], Attributes]
+  def attributesFactory: AttributesFactory
```

Comment:
We aren't reusing `AttributesFactory` between multiple implementations if I see it right. Can we keep `SpecificIterableFactory[Attribute[_], Attributes]` instead?